[IMP] l10n_es: Add taxes irnr

EmilioPascual opened the pull request:
- Add new tax groups, fiscal positions and taxes for retentions of not residents.
- Add configuration to send SII in new taxes.

https://sede.agenciatributaria.gob.es/Sede/nuevasede.html

@rafaelbn @arantxasudon

@moduon MT-5974

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/166499).

pedrobaeza commented:
This is a special case that most of the companies don't have, so not sure if it should be included in the base, or keep it as it's right now as a module in https://github.com/OCA/l10n-spain/tree/16.0/l10n_es_irnr

EmilioPascual commented:
> This is a special case that most of the companies don't have, so not sure if it should be included in the base, or keep it as it's right now as a module in https://github.com/OCA/l10n-spain/tree/16.0/l10n_es_irnr

Yes, there are companies that may not use these taxes, but that does not mean that odoo core should not have them.

Also if you use odoo core module `l10n_es_edi_sii` for send taxes to SII then you will not have these taxes configured for this module.

pedrobaeza commented:
Let's see what @jco-odoo thinks. The ideal one is that you select these cases when installing the CoA. What I want to avoid is to have the regular CoA for all the companies with 200 taxes.

jco-odoo commented:
@pedrobaeza With 200 taxes you mean in general?  It is true that the Spain CoA already has lots of them, but adding 4 here would not be too much different than what we are already doing.

pedrobaeza commented:
AFAIK, they are 8 (but the diff applying a linter makes this unreviewable) - I'm basing these numbers on https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_irnr/data/taxes_irnr.xml, plus the fiscal positions, tax groups, etc. As said, this is a very exceptional case, so adding it for all the Odoo users seems a bit bloated.

pedrobaeza commented:
Oh, I didn't know that option...

That ones then for sure deactivated by default. We can study for others in similar situation. cc @HaraldPanten

HaraldPanten commented:
> Oh, I didn't know that option...
> 
> That ones then for sure deactivated by default. We can study for others in similar situation. cc @HaraldPanten

IMHO and considering both points of the conversation:

1- Would I Add these taxes in the standard COA in Odoo? 

No, I wouldn't, because it's a non common use case for most of the companies. Because of that we have them in a separate module in the OCA.

On the other hand, Emilio's point is interesting, as if users are using SII module from Odoo core, that would be a good reason to include them in the standard COA.

2- In order to find a solution for both sides I think that we could:

- Decide which taxes should be active/inactive by default in the Spanish COA.
- Add IRNR taxes in Odoo core, as Emilio is suggesting (but keeping them inactive by default).

I think this solution would satisfy all points of view.

EmilioPascual commented:
I have just added the missing taxes referred by @pedrobaeza in https://github.com/odoo/odoo/pull/166499#issuecomment-2127091130

Also, I have updated taxes to inactive by default like @HaraldPanten commented

> 2- In order to find a solution for both sides I think that we could:
> 
> Decide which taxes should be active/inactive by default in the Spanish COA.
> Add IRNR taxes in Odoo core, as Emilio is suggesting (but keeping them inactive by default).
> I think this solution would satisfy all points of view.

It's a good solution for taxes but fiscal position can't be archived.

rafaelbn commented:
Hello @jco-odoo, @HaraldPanten , @pedrobaeza @EmilioPascual and all

This you talk about are relative and opinions, because depends on the clients.

1. 0% of our clients use Agriculture, Livestock farming, Fishing TAXES and they are there for years annoying
2. 60% of our clients use IRNR and all using SII from Odoo CE (not OCA)

Which is the correcto point of view?

All correct, all incorrect.

But HOLDED another accounting program have this taxes, Odoo is less? 😄 

We had to do a glue-module for OCA SII: https://github.com/OCA/l10n-spain/tree/15.0/l10n_es_irnr_sii

But now using Odoo SII CE, Odoo need this taxes for this use cases that are more common than you talk about.

Why archive? We have to archive always half of taxes al least, which is the problem of archaiving to more?

I agree to add this taxes not archived for everyone as other non-common taxes.

Please @EmilioPascual in any case open a OPW, Odoo should decide to support this or not for Spain anyway.

Thank you!

By the way I asked ChatGPT 😄 

> 
> 
> In Spain, the most common tax would depend on the context and who is being taxed. The Impuesto sobre la Renta de No Residentes (IRNR), or Non-Resident Income Tax, is specifically for individuals and companies who are not tax residents in Spain but earn income in the country. This could include income from property rental, business activities, or work performed in Spain.
> 
> On the other hand, taxes related to agriculture are more specialized and would typically apply to those involved in farming and agricultural businesses. These could include various forms of direct and indirect taxation, such as value-added tax (VAT) on agricultural goods, property taxes on farmland, and income tax on earnings from agricultural activities.
> 
> For the general population, IRNR would be less commonly encountered unless they are earning income in Spain without being tax residents. Conversely, taxes related to agriculture would primarily concern farmers and entities engaged in agricultural production.
> 
> Therefore, IRNR is generally more applicable to a broader group of non-residents earning income in Spain, while agricultural taxes are specific to the agricultural sector.

HaraldPanten commented:
Hi @rafaelbn If the best decision is to mark some taxes as inactive by default, I would vote to mark agriculture taxes as inactive, as well.

I agree that depending on your specialization, the taxes ara more/less common.

pedrobaeza commented:
@rafaelbn why do you partially read the threads and answer without knowing all the context? The option of having inactive taxes by default (unknown by most of us until this comment https://github.com/odoo/odoo/pull/166499#issuecomment-2127413655) changed the discussion, and now we are not debating about introducing the taxes or not, which is clear that they are going to be introduced.

Please don't add noise in the threads putting a lot of unneeded text and things that are not in the table in the moment.

etobella commented:
Can you remove the linting fixes? It makes it impossible to follow....

EmilioPascual commented:
> Can you remove the linting fixes? It makes it impossible to follow....

Done.

EmilioPascual commented:
What about this? I think it is important to have these taxes in odoo.
@jco-odoo

rafaelbn commented:
Hello @jco-odoo @chklop this taxes are needed for normal use in some cases without sii and with sii (taxes for non-resident)

Explained: 
- https://github.com/odoo/odoo/pull/166499#issue-2312544312
- https://github.com/odoo/odoo/pull/166499#issuecomment-2134067986

We developed in the past https://github.com/OCA/l10n-spain/tree/16.0/l10n_es_irnr_sii to make OCA-SII works but now we are using https://github.com/odoo/odoo/tree/17.0/addons/l10n_es_edi_sii

We would like to introduce this taxes common in our clients in order to prevent a develop like

`l10n_es_edi_sii_irnr` which should depends in a OCA module l10n_es_irnr just for some taxes.

Thank you!
Best regards
Rafael

jco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4190,6 +4190,339 @@
                 'account_id': ref('l10n_es.account_common_4751'),
                 'tag_ids': [ref('mod_111_03')],
             }),
-        ]"/>
```
be careful here

jco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4190,6 +4190,339 @@
+    <record id="account_tax_template_s_irpfnrnue0" model="account.tax.template">
+        <field name="description">IRPF no-UE exento</field>
+        <field name="type_tax_use">sale</field>
+        <field
+            name="name"
+        >Retenciones a cuenta IRPF No residentes no-UE exento por convenio</field>
```
```suggestion
        <field name="name">Retenciones a cuenta IRPF No residentes no-UE exento por convenio</field>
```

jco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4190,6 +4190,339 @@
+        <field name="amount" eval="-19" />
+        <field name="amount_type">percent</field>
+        <field name="tax_group_id" ref="tax_group_irnr_ue" />
+        <field
+            name="invoice_repartition_line_ids"
```
```suggestion
        <field name="invoice_repartition_line_ids"
```

jco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4190,6 +4190,339 @@
+        <field name="amount" eval="-24" />
+        <field name="amount_type">percent</field>
+        <field name="tax_group_id" ref="tax_group_irnr_no_ue" />
+        <field
```
It would not be bad to follow the same indentation as elsewhere: 
<field name="invoice_repartition_line_ids" eval="[(5, 0, 0),

jco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4190,6 +4190,339 @@
+        <field name="type_tax_use">sale</field>
+        <field name="name">Retenciones a cuenta IRPF No residentes UE 19%</field>
+        <field name="active">False</field>
+        <field name="chart_template_id" ref="l10n_es.account_chart_template_common" />
```
```suggestion
        <field name="chart_template_id" ref="l10n_es.account_chart_template_common"/>
```
Some linters do that, but not in the XML in Odoo.

jco-odoo commented on addons/l10n_es/data/account_tax_group_data.xml:
```diff
@@ -101,5 +101,13 @@
             <field name="country_id" ref="base.es"/>
         </record>
+        <record id="tax_group_irnr_no_ue" model="account.tax.group">
+            <field name="name">Retenciones no residentes no-UE</field>
```
Why do not we put them in the same tax groups as the once with the percentages?  Tax groups are mainly according to percentages...  They will deduce from elsewhere that it is for no residentes no-UE.

EmilioPascual replied:
I put new taxes in the same tax groups as the once with the percentages, but I had to create new tax group *Retenciones 0%*

jco-odoo commented on addons/l10n_es/data/account_fiscal_position_template_data.xml:
```diff
@@ -3862,6 +3862,701 @@
+            <field name="tax_src_id" ref="l10n_es.account_tax_template_s_iva21b" />
+            <field name="tax_dest_id" ref="l10n_es.account_tax_template_s_iva21b" />
+        </record>
+        <record
+            id="fptt_irpfnrnue24sale_21b_2"
+            model="account.fiscal.position.tax.template"
+        >
```
```suggestion
        <record id="fptt_irpfnrnue24sale_21b_2" model="account.fiscal.position.tax.template">
```

EmilioPascual commented:
@jco-odoo I attended all requests, please review again. 
Thank you.

jco-odoo commented:
@EmilioPascual https://runbot.odoo.com/runbot/build/66941787  Somehow the fiscal position itself was not defined.

jco-odoo commented:
@EmilioPascual  @rafaelbn We still do it from our side?

rafaelbn commented:
Hello @jco-odoo , yes, that's correct. However, Emilio is currently unavailable. So I ping @yajo to finish MT-5974. Thank you!

yajo commented:
Rebased & added missing fiscal positions.
